src/operator/tensor/digitize_op.h

```diff
@@ -0,0 +1,196 @@
+
+  return true;
+}
+
+
+inline bool DigitizeOpType(const nnvm::NodeAttrs &attrs,
+                           std::vector<int> *in_attrs,
+                           std::vector<int> *out_attrs) {
+  auto &data_type = (*in_attrs)[0];
+  auto &bins_type = (*in_attrs)[1];
+
+  CHECK_NE(data_type, -1) << "Input data type undefined";
+  CHECK_NE(bins_type, -1) << "Bins type undefined";
+
+  // Verify Have bins & data share the same type to simplify templating
+  CHECK_EQ(data_type, bins_type);
+
+  // Assign output_type the param
+  const int OType = nnvm::get<DigitizeParam>(attrs.parsed).otype;
+  if (OType == -1) { return false; }
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, OType);
+
+  return true;
+}
+
+
+template<typename xpu>
+struct ForwardKernel {
+  template<typename DType, typename OType>
+  static MSHADOW_XINLINE void Map(int i,
+                                  DType *in_data,
+                                  OType *out_data,
+                                  DType *bins,
+                                  size_t batch_size,
+                                  size_t bins_length,
+                                  bool right);
+};
+
+
+template<>
+struct ForwardKernel<cpu> {
+  template<typename DType, typename OType>
+  static MSHADOW_XINLINE void Map(int i,
+                                  DType *in_data,
+                                  OType *out_data,
+                                  DType *bins,
+                                  size_t batch_size,
+                                  size_t bins_length,
+                                  bool right) {
+
+    const auto data = in_data[i];
+    const auto batch = i / batch_size;
+
+    auto elem = right ? std::lower_bound(bins + bins_length * batch,
+                                         bins + bins_length * (batch + 1),
+                                         data)
+                      : std::upper_bound(bins + bins_length * batch,
+                                         bins + bins_length * (batch + 1),
+                                         data);
+
+    auto index = std::distance(bins, elem);
+    out_data[i] = OType(index);
+  }
+};
+
+
+template<typename DType>
+struct CheckMonotonic {
+  static MSHADOW_XINLINE void Map(int i, int bins_length, DType *bins) {
+    if ((i + 1) % bins_length != 0) {
+      CHECK_LT(bins[i], bins[i + 1]) << "Bins vector is not strictly monotonic and increasing";
+    } // TODO: Make sure the next element in bins is actually bins[i+1]
+  }
+};
+
+
+template<typename xpu>
+void DigitizeOpForward(const nnvm::NodeAttrs &attrs,
+                       const OpContext &ctx,
+                       const std::vector<TBlob> &inputs,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &outputs) {
+  using namespace mshadow;
+
+  auto s = ctx.get_stream<xpu>();
+  const bool right = nnvm::get<DigitizeParam>(attrs.parsed).right;
+  const auto &data = inputs[0];
```

Comment:
Too much auto!

Reply:
There is no such thing as too much auto!